[GCS] reply to the owner only after the actor has been successfully created.

wumuzi520 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?
- Problem
  - GCS Service responds immediately to the caller of the actor creation task before the actor has been created, which will cause some some problems with the reference count.

<!-- Please give a short summary of the change and the problem this solves. -->
- Solution
  - Reply to the owner only after the actor has been successfully created.

## Related issue number
Closes #8076 
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/24881/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/24883/
Test PASSed.

ericl commented:
Rllib test looks good. Is it possible to add a unit test covering this case (or python test) that doesn't need to go through all of rllib?

wumuzi520 commented:
> Rllib test looks good. Is it possible to add a unit test covering this case (or python test) that doesn't need to go through all of rllib?

@ericl  I am not quite familiar with Rllib, and do not known how to test the functionality of this PR from end to end. But I adjusted the behavior of `gcs_actor_manager_test` to make sure that the actor was created successfully before returning.
@stephanie-wang I would appreciate it if you could provide similar end-to-end testing.

BTW, I've seen some unit tests not passed. I don't think it is introduced from this PR. I fixed one of them( #8081 )

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/24893/
Test PASSed.

stephanie-wang commented:
We can add a python unit test for ref counting for actor creation. Let me try doing that and I'll push to your branch.

stephanie-wang commented:
@wumuzi520, can you give me push access to this PR? Or alternatively, here is a patch you can apply:
```
diff --git a/python/ray/tests/test_reference_counting.py b/python/ray/tests/test_reference_counting.py
index ef3462bb1..955f4c7d3 100644
--- a/python/ray/tests/test_reference_counting.py
+++ b/python/ray/tests/test_reference_counting.py
@@ -162,6 +162,29 @@ def test_dependency_refcounts(ray_start_regular):
     check_refcounts({})
 
 
+def test_actor_creation_task(ray_start_regular):
+    @ray.remote
+    def large_object():
+        # This will be spilled to plasma.
+        return np.zeros(10 * 1024 * 1024, dtype=np.uint8)
+
+    @ray.remote(resources={"init": 1})
+    class Actor:
+        def __init__(self, dependency):
+            return
+
+        def ping(self):
+            return
+
+    a = Actor.remote(large_object.remote())
+    ping = a.ping.remote()
+    ready, unready = ray.wait([ping], timeout=1)
+    assert not ready
+
+    ray.experimental.set_resource("init", 1)
+    ray.get(ping)
+
+
 def test_basic_pinning(one_worker_100MiB):
     @ray.remote
     def f(array):
```

wumuzi520 commented:
> access
@stephanie-wang 
I searched every where and still don't know how to set the permission of this PR. I directly applied your patch you mentioned above.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/24926/
Test PASSed.

stephanie-wang commented:
> > access
> > @stephanie-wang
> > I searched every where and still don't know how to set the permission of this PR. I directly applied your patch you mentioned above.

Thanks! For future reference, I think you can use this: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork
